### PR TITLE
translation: Update hash_collision.md

### PR DESCRIPTION
**Pull request by Yucao-cy**
If this pull request (PR) pertains to **Chinese-to-English translation**, please confirm that you have read the contribution guidelines and complete the checklist below:

- [x] This PR represents the translation of a single, complete document, or contains only bug fixes.
- [x] The translation accurately conveys the original meaning and intent of the Chinese version. If deviations exist, I have provided explanatory comments to clarify the reasons.

### CODE REVIEW

**Comment by krahets** [comment]
@K3v123 @thomasq0 Please review this PR at your convenience, thx!

File: docs-en/chapter_hashing/hash_collision.md [hunk 1]
```diff
@@ -1,45 +1,45 @@
...
 
-Hash collisions can lead to incorrect query results, severely affecting the usability of hash tables. To solve this problem, we expand the hash table whenever a hash collision occurs, until the collision is resolved. This method is simple and effective but inefficient due to the extensive data transfer and hash value computation involved in resizing the hash table. To improve efficiency, we can adopt the following strategies:
+Hash collisions can lead to incorrect query results, severely impacting the usability of the hash table. To address this issue, whenever a hash collision occurs, we perform hash table resizing until the collision disappears. This method is simple, straightforward, and effective, but its efficiency is too low because hash table resizing requires a large amount of data movement and hash value calculation. To improve efficiency, we can adopt the following strategies:
```
Comment (thomasq0) [inline docs-en/chapter_hashing/hash_collision.md:5]: "we **could** expand hash table" seems more straightforward. I prefer "expand" instead of "resize" because "resize" could man reducing the size.

Comment (thomasq0) [inline docs-en/chapter_hashing/hash_collision.md:5]: Recommend changing to "This approach is pretty simple, straightforward, and working well." I would argue the usage of "effective" here as the following sentence indicates this approach is ineffective. The next sentence can start with "However, "

Comment (thomasq0) [inline docs-en/chapter_hashing/hash_collision.md:5]: Suggested revision: "However, it appears to be pretty inefficient as the table expansion involves a lot of data migration as well as recalculation of hash code, which are expansive."

File: docs-en/chapter_hashing/hash_collision.md [hunk 2]
```diff
@@ -1,45 +1,45 @@
...
 
-1. Improve the data structure of the hash table, **allowing it to function normally in the event of a hash collision**.
-2. Only perform resizing when necessary, i.e., when hash collisions are severe.
+1. Improve the hash table data structure so that, **the hash table can work normally when hash collisions occur**.
```
Comment (thomasq0) [inline docs-en/chapter_hashing/hash_collision.md:7]: It is a bit vague here regarding "work normally". How about "Improve the hash table data structure in a way that **locating target element is still functioning well in the event of a hash collision**."

File: docs-en/chapter_hashing/hash_collision.md [hunk 3]
```diff
@@ -1,45 +1,45 @@
...
+2. Only perform resizing when necessary, i.e., when hash collisions are relatively severe.
```
Comment (thomasq0) [inline docs-en/chapter_hashing/hash_collision.md:8]: rewrite to "Expansion is the last resort before it becomes necessary, when severe collisions are observed."

**Review by thomasq0 — CHANGES_REQUESTED** [review]

File: docs-en/chapter_hashing/hash_collision.md [hunk 4]
```diff
@@ -1,45 +1,45 @@
 # Hash Collision
 
-As mentioned in the previous section, **usually the input space of a hash function is much larger than its output space**, making hash collisions theoretically inevitable. For example, if the input space consists of all integers and the output space is the size of the array capacity, multiple integers will inevitably map to the same bucket index.
+The previous section mentioned that, **in most cases, the input space of a hash function is much larger than the output space**, so theoretically, hash collisions are inevitable. For example, if the input space is all integers and the output space is the size of the array capacity, then multiple integers will inevitably be mapped to the same bucket index.
```
Comment (K3v123) [inline docs-en/chapter_hashing/hash_collision.md:3]: um it might flow better if you change: "The previous section mentioned that:" to -> "In the previous section, it was mentioned that."
or just stick with original. -> "As mentioned in the previous section"

"in most cases, the input space of a hash function is much larger than the output" -> "than its output"

"For example, if the input space is" -> "For example, if the input space consists of" (original flows better) or you can do: ->  "if the input space comprises all integers and the output space is the size"


optional "in most cases" -> usually (like original) or typically. but this is fine anyways, its just a bit wordy ish.
"so theoretically, hash collisions are inevitable." -> "hence making hash collisions theoretically unavoidable."

File: docs-en/chapter_hashing/hash_collision.md [hunk 5]
```diff
@@ -1,45 +1,45 @@
...
+In the original hash table, each bucket can store only one key-value pair. "Separate chaining" converts a single element into a linked list, treating key-value pairs as list nodes, storing all colliding key-value pairs in the same linked list. The figure below shows an example of a hash table with separate chaining.
 
 ![Separate Chaining Hash Table](hash_collision.assets/hash_table_chaining.png)
 
 The operations of a hash table implemented with separate chaining have changed as follows:
 
-- **Querying Elements**: Input `key`, pass through the hash function to obtain the bucket index, access the head node of the list, then traverse the list and compare `key` to find the target key-value pair.
+- **Querying Elements**: Input `key`, obtain the bucket index through the hash function, then access the head node of the linked list, traverse the linked list, and compare `key` to find the target key-value pair.
```
Comment (K3v123) [inline docs-en/chapter_hashing/hash_collision.md:20]: "linked list, traverse the linked list, and compare `key` to find the target key-value pair." -> "linked list. Traverse the linked list and compare key to find the target key-value pair."

File: docs-en/chapter_hashing/hash_collision.md [hunk 6]
```diff
@@ -1,45 +1,45 @@
...
 - **Adding Elements**: First access the list head node via the hash function, then add the node (key-value pair) to the list.
```
Comment (K3v123) [inline docs-en/chapter_hashing/hash_collision.md:21]: optional for more clarity: "access the head node of the linked list via the hash function, then append the node (key-value pair) to the list."

File: docs-en/chapter_hashing/hash_collision.md [hunk 7]
```diff
@@ -1,45 +1,45 @@
...
-- **Deleting Elements**: Access the list head based on the hash function's result, then traverse the list to find and remove the target node.
+- **Deleting Elements**: Access the head of the linked list based on the result of the hash function, then traverse the linked list to find the target node and delete it.
 
 Separate chaining has the following limitations:
 
 - **Increased Space Usage**: The linked list contains node pointers, which consume more memory space than arrays.
-- **Reduced Query Efficiency**: Due to the need for linear traversal of the list to find the corresponding element.
+- **Reduced Query Efficiency**: Because linear traversal of the linked list is required to find the corresponding element.
```
Comment (K3v123) [inline docs-en/chapter_hashing/hash_collision.md:27]: um optional: "Because linear traversal of th" -> "This is because linear traversal"

File: docs-en/chapter_hashing/hash_collision.md [hunk 8]
```diff
@@ -1,45 +1,45 @@
...
 
 The code below provides a simple implementation of a separate chaining hash table, with two things to note:
 
 - Lists (dynamic arrays) are used instead of linked lists for simplicity. In this setup, the hash table (array) contains multiple buckets, each of which is a list.
-- This implementation includes a method for resizing the hash table. When the load factor exceeds $\frac{2}{3}$, we resize the hash table to twice its original size.
+- This implementation includes a hash table resizing method. When the load factor exceeds $\frac{2}{3}$, we resize the hash table to double its original size.
```
Comment (K3v123) [inline docs-en/chapter_hashing/hash_collision.md:32]: i think "twice" is better (flows better)

**Review by K3v123 — COMMENTED** [review]
thank you for the PR and your hard work Yucao-cy!

its really good. 
thanks

**Comment by krahets** [comment]
I was moving the `docs-en` to `en/docs` in the main branch and the changes had been merged into this PR. Please synchronize the latest commits to your local repo. Sorry for inconvenice.

**Review by K3v123 — APPROVED** [review]

File: docs-en/chapter_hashing/hash_collision.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (krahets) [reply docs-en/chapter_hashing/hash_collision.md]: In Chinese, we use "输入空间" and "输出空间" to mathematically represent the space of the input and output. I think we can keep "input space" and "input space" if it is authentic in English.

[the remark this answers is not in the document]
Comment (krahets) [reply docs-en/chapter_hashing/hash_collision.md]: The CN version is `上一节提到，通常情况下哈希函数的输入空间远大于输出空间，因此理论上哈希冲突是不可避免的。` FYI

File: docs-en/chapter_hashing/hash_collision.md (hunk 2 above)
Comment (krahets) [reply docs-en/chapter_hashing/hash_collision.md]: This looks more accurate! I will finetune the Chinese version too though. Please @ me for future changes like this one, thanks!

Comment (krahets) [reply docs-en/chapter_hashing/hash_collision.md]: The Chinese version: `1. 改良哈希表数据结构，**使得哈希表可以在出现哈希冲突时正常工作**。`
We've introduced the details of "非正常工作" caused by hash collision in the previous section. 

Considering the context, do you think "Hash tables can still work correctly." is vague? If not, I prefer the concise one.

File: en/docs/chapter_hashing/hash_collision.md [hunk 9]
```diff
@@ -1,108 +1,111 @@
...
-2. Only perform resizing when necessary, i.e., when hash collisions are severe.
+1. Improve the hash table data structure in a way that **locating target element is still functioning well in the event of a hash collision**.
+2. Expansion is the last resort before it becomes necessary, when severe collisions are observed.
 
 There are mainly two methods for improving the structure of hash tables: "Separate Chaining" and "Open Addressing".
 
 ## Separate chaining
 
-In the original hash table, each bucket can store only one key-value pair. "Separate chaining" transforms individual elements into a linked list, with key-value pairs as list nodes, storing all colliding key-value pairs in the same list. The figure below shows an example of a hash table with separate chaining.
+In the original hash table, each bucket can store only one key-value pair. "Separate chaining" converts a single element into a linked list, treating key-value pairs as list nodes, storing all colliding key-value pairs in the same linked list. The figure below shows an example of a hash table with separate chaining.
```
Comment (thomasq0) [inline en/docs/chapter_hashing/hash_collision.md:14]: add "and " before "storing all colliding key-value pairs in the same linked list."

File: en/docs/chapter_hashing/hash_collision.md [hunk 10]
```diff
@@ -1,108 +1,111 @@
...
 
 ![Separate chaining hash table](hash_collision.assets/hash_table_chaining.png)
 
 The operations of a hash table implemented with separate chaining have changed as follows:
 
-- **Querying elements**: Input `key`, pass through the hash function to obtain the bucket index, access the head node of the list, then traverse the list and compare `key` to find the target key-value pair.
-- **Adding elements**: First access the list head node via the hash function, then add the node (key-value pair) to the list.
-- **Deleting elements**: Access the list head based on the hash function's result, then traverse the list to find and remove the target node.
+- **Querying Elements**: Input `key`, obtain the bucket index through the hash function, then access the head node of the linked list. Traverse the linked list and compare key to find the target key-value pair.
```
Comment (thomasq0) [inline en/docs/chapter_hashing/hash_collision.md:20]: I believe we have adopted the wiki style heading where only the first word is capitalized.

Comment (thomasq0) [reply en/docs/chapter_hashing/hash_collision.md:20]: Rephrase a bit using imperative sentence because `key` is the input rather than part of the flow:

"Given a `key` input, determine the bucket index using the hash function, ..."

File: en/docs/chapter_hashing/hash_collision.md [hunk 11]
```diff
@@ -1,108 +1,111 @@
...
 [file]{hash_map_chaining}-[class]{hash_map_chaining}-[func]{}
 ```
 
-It's worth noting that when the list is very long, the query efficiency $O(n)$ is poor. **At this point, the list can be converted to an "AVL tree" or "Red-Black tree"** to optimize the time complexity of the query operation to $O(\log n)$.
+It's worth noting that when the linked list is very long, the query efficiency $O(n)$ is poor. **In this case, the list can be converted to an "AVL tree" or "Red-Black tree"** to optimize the time complexity of the query operation to $O(\log n)$.
 
 ## Open addressing
 
-"Open addressing" does not introduce additional data structures but uses "multiple probes" to handle hash collisions. The probing methods mainly include linear probing, quadratic probing, and double hashing.
+"Open addressing" does not introduce additional data structures but instead handles hash collisions through "multiple probing". The probing methods mainly include linear probing, quadratic probing, and double hashing.
```
Comment (thomasq0) [inline en/docs/chapter_hashing/hash_collision.md:43]: We do not use "but" and "instead" together. I would suggest changing the second part to ", instead it handles ...".

File: en/docs/chapter_hashing/hash_collision.md [hunk 12]
```diff
@@ -1,108 +1,111 @@
...
 
-However, **linear probing tends to create "clustering"**. Specifically, the longer a continuous position in the array is occupied, the more likely these positions are to encounter hash collisions, further promoting the growth of these clusters and eventually leading to deterioration in the efficiency of operations.
+However, **linear probing is prone to create "clustering"**. Specifically, the longer the continuously occupied positions in the array, the greater the probability of hash collisions occurring in these continuous positions, further promoting the growth of clustering at that position, forming a vicious cycle, and ultimately leading to degraded efficiency of insertion, deletion, query, and update operations.
 
 It's important to note that **we cannot directly delete elements in an open addressing hash table**. Deleting an element creates an empty bucket `None` in the array. When searching for elements, if linear probing encounters this empty bucket, it will return, making the elements below this bucket inaccessible. The program may incorrectly assume these elements do not exist, as shown in the figure below.
```
Comment (thomasq0) [inline en/docs/chapter_hashing/hash_collision.md:61]: update 2nd sentence to: "This is because deleting an element would leave an empty bucket `None` in the array."

Comment (thomasq0) [reply en/docs/chapter_hashing/hash_collision.md:61]: update 3rd sentence to: "When searching for elements, linear probing would stop at this empty bucket, causing the rest of elements below it inaccessible."

File: docs-en/chapter_hashing/hash_collision.md (hunk 2 above)
Comment (thomasq0) [reply docs-en/chapter_hashing/hash_collision.md]: agree that it is not vague given that previous session gives the context about what might go wrong. We can simplify to "the hash table can function well when hash collisions occur".

**Review by krahets — APPROVED** [review]
Thanks for the PR!
